### ISSUE TITLE
sway: Optimal NixOS integration

### ISSUE THREAD

**Issue by primeos** [NixOS/nixpkgs#57602]
This is intended as an issue to collect and track possible improvements and feedback in general. Feedback form users is welcome (what is difficult, does not work, could be improved, are packages missing, etc.).

Possible improvements (WIP):
- Should we install the default wallpaper? (see https://github.com/NixOS/nixpkgs/pull/57414#discussion_r265056873)
- systemd integration (user units)
- Login manager integration

**Comment by jtojnar** [comment]
virtboard is packaged in https://github.com/NixOS/nixpkgs/pull/57614

**Comment by gnidorah** [comment]
> light / brightnessctl: backlight management

Could actkbd or acpid used instead? For brightness control locally I use
```
  brightness = pkgs.writeShellScriptBin "brightness.sh" ''
    bl_dev=`echo /sys/class/backlight/*`
    echo $(($(< $bl_dev/brightness) $1 $2)) >$bl_dev/brightness
  '';
```
Will this work for all video cards? If so, I would like to add brightness control defaults to actkbd, like we already have for sound control `sound.mediaKeys.enable = true;`

**Comment by gnidorah** [comment]
> Useful software (for Wayland in general) in nixpkgs:
> wallutils

Don't see it in nixpkgs.

> systemd integration (user units) - WIP: #54627 (comment)

What benefits it brings? sway is capable to run applications on start by itself.

**Comment by t184256** [comment]
> What benefits it brings?

Starting sway itself, presumably.

**Comment by primeos** [comment]
> Could actkbd or acpid used instead?

I do not have any preference here, it was simply meant as a list of software that could be useful (I took if from [Drew's blog post](https://drewdevault.com/2019/03/11/Sway-1.0-released.html)). I also didn't plan to integrate this into the `sway` module if that was your concern (not sure if I'm interpreting your comment correctly).

> If so, I would like to add brightness control defaults to actkbd, like we already have for sound control `sound.mediaKeys.enable = true;`

Might be a good idea if we already have this for sound. But since this is a generic solution that should work with both Wayland and X11 we should not put it in the `sway` module (but that was probably not your intention anyway).

> Don't see it in nixpkgs.

Only the checked entries (checkbox) are already in nixpkgs.

> What benefits it brings? sway is capable to run applications on start by itself.

I'm not using user units - it's mainly useful for people who already use them. I assume the main advantages are to automatically launch sway on a TTY (like with `.bashrc`), "dependency management" (other systemd user units) and we could better integrate it into nixpkgs than via the sway configuration (though I'm not sure how useful it would actually be - but it could be implemented in a generic and standardized way I guess).

**Comment by amazari** [comment]
Home-manager provided a nice way to set i3 declaratively [0].
Could be nice to have that in-tree and shared with i3 or add sway support to home-manager's dingy.

[0] https://github.com/rycee/home-manager/blob/master/modules/services/window-managers/i3.nix

**Comment by gnidorah** [comment]
@t184256 
> Starting sway itself, presumably.

Looks like we have a blocker for user services: https://github.com/NixOS/nixpkgs/issues/21460

@primeos 
> But since this is a generic solution that should work with both Wayland and X11 we should not put it in the sway module (but that was probably not your intention anyway).

Right, I'm just interesting whether it will work with different video cards. I only have Intel there.

> Only the checked entries (checkbox) are already in nixpkgs.

Sorry. With sway 1.0-rc5, firefox and breeze dark theme they are not visible by some reason.

**Comment by CrazedProgrammer** [comment]
wf-recorder has been merged! #57662  :tada:

**Comment by primeos** [comment]
> Right, I'm just interesting whether it will work with different video cards. I only have Intel there.

I also only have Intel but according to the [Arch Wiki](https://wiki.archlinux.org/index.php/backlight):
> All methods are exposed to the user through /sys/class/backlight

it sounds like a good method that should work in most cases.

But we would have to modify the script to consider [`max_brightness`](https://www.kernel.org/doc/Documentation/ABI/stable/sysfs-class-backlight).

According to the Arch Wiki:
> The maximum brightness can be found by reading from max_brightness, which is often `15`.

and on my system `max_brightness` has the value `7500`. So the step size of `10` is either too high or too low in these cases. Using something like `10%` would be better.

> With sway 1.0-rc5, firefox and breeze dark theme they are not visible by some reason.

As long as it works with 1.0 :)

> Home-manager provided a nice way to set i3 declaratively [0].
Could be nice to have that in-tree and shared with i3 or add sway support to home-manager's dingy.

@amazari interesting. I don't think that we can/should move it in-tree (too many configuration options IMO) but I'm assuming that adding a module for `sway` to home-manager would be fine - any volunteers? :)

**Comment by wkral** [comment]
Waybar has been merged. #54627

**Comment by gnidorah** [comment]
@primeos Would be nice to add environment setup like that https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix does
For e.g. currently without doing 
`exec dbus-update-activation-environment DISPLAY`
in sway config I get an error when trying to send notification
`gi.repository.GLib.Error: g-dbus-error-quark: Ошибка вызова StartServiceByName для org.freedesktop.Notifications: GDBus.Error:org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.Notifications exited with status 1 (25)`

**Comment by NilsIrl** [comment]
What about [bemenu](https://github.com/Cloudef/bemenu)?

**Comment by primeos** [comment]
@NilsIrl good idea, I added it to the list - it's already packaged in nixpkgs ([pkgs/applications/misc/bemenu/default.nix](https://github.com/NixOS/nixpkgs/blob/83a01ab93a225a230c848f6038bc0dc6a6b8f217/pkgs/applications/misc/bemenu/default.nix))

@gnidorah seems like I forgot to reply :o

Regarding the environment setup: Yes, that would be nice. I think there is already an issue to unify this for X11 and Wayland (I'll try to look it up and reference it).

Regarding the notification: Can this be reproduced (e.g. with `notify-send` and `mako`)? I don't use notifications but I could have a look at it if it doesn't work.

**Comment by gnidorah** [comment]
@primeos 
No, mako works. Looks like this is only needed for X11 programs. Thanks, I will switch to mako from now.
Though with mako I see an issue that it can't be launched automatically
```
gi.repository.GLib.Error: g-dbus-error-quark: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.Notifications was not provided by any .service files (2)
```
If I start mako manually, everything works fine

**Comment by primeos** [comment]
@gnidorah great :) Just realized that I missed the `exec dbus-update-activation-environment DISPLAY` part...

---

```
gi.repository.GLib.Error: g-dbus-error-quark: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.Notifications was not provided by any .service files (2)
```

Seems like `mako` didn't provide a .service file but it should work after the next release:
https://github.com/emersion/mako/commit/ca8e763f06756136c534b1bbd2e5b536be6b1995 (cc @dywedir maybe you could you test if it works after the next release)

**Comment by gnidorah** [comment]
@NilsIrl

How to make it work under sway? This doesn't show anything for me
```
nix-build -A bemenu
./result/bin/bemenu
```

**Comment by NilsIrl** [comment]
> @NilsIrl
> 
> How to make it work under sway? This doesn't show anything for me
> 
> ```
> nix-build -A bemenu
> ./result/bin/bemenu
> ```

I had the same problem. Bemenu is only in the master branch.
```bash
git clone git@github.com:NixOS/nixpkgs.git
nix-build -A bemenu -I nixpkgs=nixpkgs/
```

**Comment by gnidorah** [comment]
@NilsIrl I've tried it from master branch, but when I run `bemenu` under sway it doesn't draw anything on the screen.

**Comment by NilsIrl** [comment]
> @NilsIrl I've tried it from master branch, but when I run `bemenu` under sway it doesn't draw anything on the screen.

Have you tried to run bemenu from a terminal?

**Comment by gnidorah** [comment]
> Have you tried to run bemenu from a terminal?

That is. And nothing on the screen. Sway version is 1.1.1

**Comment by NilsIrl** [comment]
> > Have you tried to run bemenu from a terminal?
> 
> That is. And nothing on the screen. Sway version is 1.1.1

If you can open a terminal, you could see what bemenu outputs when it is run. If you don't have access to a terminal, you can try to redirect the output from bemenu into a file and then view the file.

**Comment by NilsIrl** [comment]
I personally had a problem where I couldn't even run a terminal. What I did is that I redirected stdout and stderr from the command launching the terminal in sway into a file.

I then viewed the file from the tty. And guess what... The file had an error message saying my version of opengl was too outdated for the terminal emulator, alacritty to work.

**Comment by gnidorah** [comment]
@NilsIrl Looks like my mistake was in that I've tried to run `bemenu` instead of `bemenu-run`. bemenu-run works just fine. My sorry :disappointed: 
BTW I hope they introduce something akin to `i3-dmenu-desktop` some day

**Comment by NilsIrl** [comment]
@gnidorah 
https://github.com/i3/i3/blob/6339427f017e1265a022e6537e16a8b4a921e52f/i3-dmenu-desktop
You can modify this script for it to use bemenu instead of dmenu.

**Comment by NilsIrl** [comment]
Actually you don't even need to do that.
Just run `$ i3-dmenu-desktop --demenu=bemenu` and it should work

If you don't want to install i3 just to have i3-dmenu-desktop you can download it from https://github.com/i3/i3/blob/6339427f017e1265a022e6537e16a8b4a921e52f/i3-dmenu-desktop

**Comment by matthias-t** [comment]
i'd like to suggest redshift to adjust the color temperature:
https://aur.archlinux.org/packages/redshift-wlr-gamma-control-git/

optimally there could be a switch in `services.redshift`

**Comment by matthias-t** [comment]
also, https://github.com/ReimuNotMoe/ydotool

**Comment by NilsIrl** [comment]
wshowkeys, https://git.sr.ht/~sircmpwn/wshowkeys

mentioned in https://github.com/swaywm/sway/wiki/i3-Migration-Guide

**Comment by erictapen** [comment]
Regarding systemd user service integration: I started to document my setup in https://nixos.wiki/wiki/Sway. Atm I have no time to integrate these changes into NixOS, but maybe somebody else has?

**Comment by Elyhaka** [comment]
Hello @erictapen, I've been trying to integrate Sway with systemd the way you proposed, but I had multiple issues doing so : 
 - I could not find a way to autostart correctly the session
 - I could not find a way bind it with a displaymanager properly at all

Have you been able to achieve this ?

The approach i' m currently following is to start Sway as the display-manager itself. I have the feeling the way you're doing it might be cleaner. Here is my setup :

```
  systemd.defaultUnit = "graphical.target";
  systemd.services.sway = {
    enable = true;
    wants = [ "systemd-machined.service" ];
    aliases = [ "display-manager.service" ];
    after =  [
      "rc-local.service"
      "systemd-machined.service"
      "systemd-user-sessions.service"
      "plymouth-quit.service"
      "plymouth-start.service"
    ];
    serviceConfig = {
      ExecStartPre = "${config.system.path}/bin/chvt ${sway.tty}";
      ExecStart = "${pkgs.dbus}/bin/dbus-launch --exit-with-session ${pkgs.sway}/bin/sway";
      TTYPath = "/dev/tty${sway.tty}";
      TTYReset = "yes";
      TTYVHangup = "yes";
      TTYVTDisallocate = "yes";
      PAMName = "login";
      User = sway.username;
      WorkingDirectory = "/home/${sway.username}";
      StandardInput = "tty";
      StandardError = "journal";
      StandardOutput = "journal";
      Restart = "no";
    };
  };
```

It still [has one issue](https://discourse.nixos.org/t/failed-to-create-session-unit-dbus-org-freedesktop-login1-service-not-found/4988) though.

**Comment by thinkpadder1** [comment]
> Regarding systemd user service integration: I started to document my setup in https://nixos.wiki/wiki/Sway. Atm I have no time to integrate these changes into NixOS, but maybe somebody else has?

I tried your config, but had to make some changes (due to redshift):
Remove
```
    package = pkgs.redshift-wlr;
```

Add (top level)
```
location = {
 latitude = 40.7;
 longitude = -74.0;
}
```

Also, not relevant to redshift, but
```
    environment.PATH = lib.mkForce null;
```
had to be changed to
```
    environment.PATH = pkgs.lib.mkForce null;
```
because `lib` does not exist (not sure if I did this one right).

**Comment by NilsIrl** [comment]
When shutting down, sway is inhibiting the shutdown from what I can tell.

This causes very long shutdowns.

**Comment by Bryophyllum** [comment]
I am sorry if this is not a right place to drop my idea, or this is something to create a separate issue for, but I would personally love if [the Sway package](https://github.com/NixOS/nixpkgs/blob/release-19.09/nixos/modules/programs/sway.nix#L7) in the option `programs.sway` was exposed so that it could be overridden with a package from the nixos-unstable.

Since the Wayland ecosystem and Sway are still in work in progress adoption-wise; many Wayland tools, and the newest releases of Wlroots and Sway themselves rarely make it into Linux distributions, which follow the fixed 6 months release schedule model, in time. While using NixOS Unstable may be an option, having the latest Wayland stack is all I need.

My lack of understanding of the basic knowledge about Nix and NixOS'es Sway wrapper may make obvious limitations of implementing this feature inapparent to me, but I would like to hear from you what you think about it, and why it could/should (not) be implemented, nonetheless.

https://functor.tokyo/blog/2018-02-18-install-packages-from-nixos-unstable

**Comment by primeos** [comment]
@Bryophyllum that's a valid idea and feel free to add such comments to this issue ;)

That said, I'm not sure if having such a `package` option is worth it. I don't have any strong feelings on this, but AFAIK the consensus is to avoid such an option in cases like this (see https://github.com/NixOS/nixpkgs/issues/50476). Also: Mixing Sway from nixos-unstable with a system based on nixos-stable can result im problems like https://github.com/NixOS/nixpkgs/pull/76787#issuecomment-578461880.

Therefore I would avoid adding such an option. However, you should still be able to realise your use-case by using overlays. It should be sufficient to override the sway package with the one from nixos-unstable. Here is an example: https://gist.github.com/LnL7/58d687c58948e71acd7fd4d410f3cdd1 (instead of `unstable = ...` the following should be sufficient: `sway = unstable.sway`).

**Comment by eoli3n** [comment]
would you add : https://github.com/NixOS/nixpkgs/issues/78370 ?

**Comment by primeos** [comment]
For users interested in Grimshot or other scripts in `contrib/`: In #87979 we're currently discussing how to best package these in Nixpkgs.

**Comment by flokli** [comment]
I'm still missing some documentation / configuration examples on how to actually /enable/ sway.

While testing pipewire support in a VM, I hacked together the following:

```
      programs.sway.enable = true;

      services.xserver.enable = true; #??
      services.xserver.displayManager.sddm = {
        enable = true;
        autoLogin = {
          enable = true;
          user = "alice";
        };
      };
      services.xserver.displayManager.defaultSession = "sway";

      xdg.portal = {
        enable = true;
        extraPortals = with pkgs; [ xdg-desktop-portal-wlr ];
        gtkUsePortal = true;
      };

      services.pipewire.enable = true;
```

However, this is using Xwayland, and this is probably not right, i don't know.

We should also have some tests that ensure stuff doesn't break - I tried automating something with `ydotool`, but couldn't get it to work.

**Comment by primeos** [comment]
> I'm still missing some documentation / configuration examples on how to actually /enable/ sway.

Where would you suggest to document this? E.g. there's [a page on the NixOS wiki](https://nixos.wiki/wiki/Sway) but that is currently only focused on systemd. Alternatively we could document it in the NixOS manual, but unfortunately there are many Wayland compositors and we don't have a general abstraction for them in Nixpkgs (i.e. documenting Sway there might be too specific).

But basically `programs.sway.enable = true;` is already enough and the most relevant things should already be documented via the NixOS module.

> However, this is using Xwayland, and this is probably not right, i don't know.

I guess this is regarding Chromium? Unfortunately Ozone still isn't considered stable / enabled by default so you have to recompile Chromium for native Wayland support: `chromium.override { useOzone = true; }`. You can also disable XWayland by removing it from `programs.sway.extraPackages` or via the Sway config (IIRC).

> We should also have some tests that ensure stuff doesn't break - I tried automating something with ydotool, but couldn't get it to work.

A NixOS VM test for Sway sounds like a good idea :)

**Comment by flokli** [comment]
> Where would you suggest to document this? E.g. there's [a page on the NixOS wiki](https://nixos.wiki/wiki/Sway) but that is currently only focused on systemd. Alternatively we could document it in the NixOS manual, but unfortunately there are many Wayland compositors and we don't have a general abstraction for them in Nixpkgs (i.e. documenting Sway there might be too specific).

Yeah, I'd prefer having more complicated things documented in the NixOS manual via `meta.doc` - those hopefully are more likely to be up-to-date with the code than a page in the wiki.
A Sway-specific documentation is probably better than no documentation at all :-)

Also note the [Wiki Page](https://nixos.wiki/wiki/Sway) is more a configuration dump that shows a way to get it working with systemd (and probably outdated, see https://github.com/NixOS/nixpkgs/pull/85334#issuecomment-615495925 as discussed with @Ma27)

These parts should really just end up in the sway module, and once we want to abstract that to support more wayland compositors, can be done as a follow-up.

> I guess this is regarding Chromium? Unfortunately Ozone still isn't considered stable / enabled by default so you have to recompile Chromium for native Wayland support […]

Once compiled with Ozone support, is this a flag/environment variable? Could we enable this during compilation, while still having users opting in (similar to https://github.com/NixOS/nixpkgs/pull/89565)?

> A NixOS VM test for Sway sounds like a good idea :)

As I posted in my [previous comment](https://github.com/NixOS/nixpkgs/issues/57602#issuecomment-640018631), I didn't get the `ydotool` parts to work, so we can't really do anything in it. I'd assume for the test, we'd at least want to open a terminal and do some OCR (like in `nixosTests.cage`)

**Comment by primeos** [comment]
> A Sway-specific documentation is probably better than no documentation at all :-)

Ok, I'll try to look into that then (but it might take me some time). Help is obviously welcome, if someone else who follows this issue is interested in that.

> Once compiled with Ozone support, is this a flag/environment variable?

You need to run `chromium --ozone-platform=wayland` to select the Wayland backend (by default X11 is used).

> Could we enable this during compilation, while still having users opting in (similar to #89565)?

Unfortunately not :o See https://github.com/NixOS/nixpkgs/issues/82443 and https://github.com/NixOS/nixpkgs/pull/85197 for related discussions. I'm hoping that Ozone will become stable in 2020 and current news look promising (e.g. [Chrome Is Reaching The Point Of Good X11 + Wayland Support In Same Build](https://www.phoronix.com/scan.php?page=news_item&px=Chrome-Ozone-X11-Build-Good)  - the X11 stability was lacking behind Wayland, which is the main focus).

> As I posted in my previous comment, I didn't get the ydotool parts to work, so we can't really do anything in it. I'd assume for the test, we'd at least want to open a terminal and do some OCR (like in nixosTests.cage)

Why don't we just send the key combinations via `send_keys` to launch a terminal, etc.?

**Comment by wkral** [comment]
> ```services.xserver.enable = true; #??```

I was curious about that myself as I'd rather not have everything else that brings in.

To have just a display manager that launches sway and other wayland window managers all that is needed from [`xserver.nix`](https://github.com/NixOS/nixpkgs/blob/e92b11d964d13991f5d62b3affec04c4fb03f110/nixos/modules/services/x11/xserver.nix#L649)  is setting`systemd.defaultUnit` and creating the display manager service. I pulled those out manually on my local machine to verify.

I'm wondering if extracting that config into display manager specific config might be a good idea. Maybe something where if `services.xserver.enabled` is true, then the display manager is automatically enabled, but it could be enabled without enabling the xserver.

I'd appreciate any feedback and opinions on that idea, and then I could potentially put together a PR.

**Comment by Algram** [comment]
Has anyone gotten a different login manager than sddm working with sway? I am currently trying `lightdm` with the mini greeter, but I only get a black sceen with a cursor.

Maybe something like https://git.sr.ht/~kennylevinsen/gtkgreet would be perfect since it seems to be made specifically for sway?

**Comment by wkral** [comment]
> Has anyone gotten a different login manager than sddm working with sway? I am currently trying `lightdm` with the mini greeter, but I only get a black sceen with a cursor.

I've been using `gdm` with sway, as it's the only one that I'm aware of that will render with wayland and not have to start an Xserver. I did it with the proposed method above where just the display manager is enabled without enabling all of Xorg.

In my `configuration.nix`:
```nix
  services.xserver.displayManager.gdm = {
    enable = true;
    wayland = true;
  };

  # Extracted from nixos/modules/services/x11/xserver.nix
  systemd.defaultUnit = "graphical.target";
  systemd.services.display-manager =
    let
      cfg = config.services.xserver.displayManager;
    in
    {
      description = "Display Manager";

      after = [ "acpid.service" "systemd-logind.service" ];

      restartIfChanged = false;

      environment =
        lib.optionalAttrs
          config.hardware.opengl.setLdLibraryPath {
          LD_LIBRARY_PATH = pkgs.addOpenGLRunpath.driverLink;
        } // cfg.job.environment;

      preStart =
        ''
          ${cfg.job.preStart}

          rm -f /tmp/.X0-lock
        '';

      script = "${cfg.job.execCmd}";

      serviceConfig = {
        Restart = "always";
        RestartSec = "200ms";
        SyslogIdentifier = "display-manager";
        # Stop restarting if the display manager stops (crashes) 2 times
        # in one minute. Starting X typically takes 3-4s.
        StartLimitInterval = "30s";
        StartLimitBurst = "3";
      };
    };

```

I previously checked out gtkgreet, it's one of the greeters that are part of [greetd](https://git.sr.ht/~kennylevinsen/greetd). That project seems really promising but still in early phases. Maybe some progress has been made since I last looked. Also since it involves login, hopefully there will be some security minded folks looking over it.

**Comment by berbiche** [comment]
> Has anyone gotten a different login manager than sddm working with sway? I am currently trying `lightdm` with the mini greeter, but I only get a black sceen with a cursor.

I have been using lightdm with Sway successfully on `nixpkgs-unstable` and `20.03` channel with no problems whatsoever.
I use the enso greeter if it's relevant.
My keyring (gnome-keyring) gets automatically unlocked too.

Excerpt from my config:

``` nix
{
  services.xserver.enable = true;
  services.xserver.displayManager.defaultSession = "sway";
  services.xserver.displayManager.lightdm = {
    enable = true;
    greeters.enso.enable = true;
  };
  services.xserver.libinput.enable = true;
  services.xserver.layout = "us";

  services.gnome3.gnome-keyring.enable = true;
}
```

**Comment by primeos** [comment]
I've never tried screen sharing myself but for Firefox #84233 might fix your issues and for Chromium #89565 and linked issues/PRs might contain valuable information (configuration examples, etc.).

**Comment by t184256** [comment]
Meanwhile, wf-recoder might or might not save you from relogging into XFCE.

**Comment by ddelabru** [comment]
> Meanwhile, wf-recoder might or might not save you from relogging into XFCE.

I did try an alternate configuration involving the `v4l2loopback` kernel module and wf-recorder but when I tried to initialize a "webcam" feed from wf-recorder I ran into some errors. If I recreate that configuration I could share the errors I got here. It feels like kind of a hack to begin with though; if I could get one of these solutions to work I'd prefer it to be pipewire.

**Comment by ddelabru** [comment]
> > Meanwhile, wf-recoder might or might not save you from relogging into XFCE.
> 
> If I recreate that configuration I could share the errors I got here.

Alright, I tried again to adapt some instructions from the Arch Wiki to use wf-recorder to create a virtual webcam, and (good news!) this time I succeeded. Still feels a little hacky, but this is what I did.

In my NixOS config:
```
  # Obviously omitting all my non-screen-sharing-related packages:
  programs.sway.extraPackages = with pkgs; [
    v4l2-utils
    wf-recorder
  ];
  boot.extraModulePackages = with config.boot.kernelPackages; [
    v4l2loopback
  ];
```

Then when I'm ready to start screen-sharing I do

```
# modprobe v4l2loopback exclusive_caps=1 card_label=WfRecorder
$ v4l2-ctl --list-devices
```

On my system this results in `/dev/video4` being allocated with the `WfRecorder` label, so then I do

```
$ wf-recorder --muxer=v4l2 --codec=rawvideo --file=/dev/video4 -x yuv420p
```

I had two displays active when I first tried this, so wf-recorder asked me to specify which one I wanted to capture. I left this command running while I opened Firefox, opened a webcam-sharing test, and easily shared the "webcam" that was capturing my screen.

Now that I have this working I guess it's my solution for the time being. Would have preferred to just select applications or displays from within Firefox on the fly instead of setting up a virtual webcam first, but this is a pretty cool workaround.

**Comment by mschwaig** [comment]
As someone who is relatively new to NixOS and configuring this kind of thing I still found it fairly difficult to pull together all of the required pieces from different places to get a working setup with ` sway` as the window manager and ` gdm` as the display manager (to avoid having to enable `xserver` - not sure how much sway users care about this).

I'm still having trouble with kanshi, other than that I have a working config, which I wanted to share in case others find it useful:
https://gist.github.com/mschwaig/195fe93ed85dea7aaceaf8e1fc6c0e99

I also added a warning in the wiki because I forgot to start the `sway-session.target` and this gave me trouble for a long time. If kanshi was working and I knew that GDM was a good choice for most people (SDDM is mentioned in the original issue) I'd be happy to extend the wiki with a few lines on display manager integration..

There is a [blogpost by Ongy](https://blog.ongy.net/posts/sway-and-systemd/) who pulled out more stuff than just kanshi into individual units. I don't know if that would be a good idea.

**Comment by primeos** [comment]
@mschwaig yeah, the current module is focused on running Sway from a virtual terminal so using login/display managers (especially without X11) and/or systemd user units is unfortunately a bit tricky. For the login/display managers we'd likely need better NixOS abstractions that aren't so much X11 focused (not sure how well this is currently possible though). I'm also not sure what to do about the systemd user unit integration. Upstream doesn't officially support it (e.g. [noted in the wiki](https://github.com/swaywm/sway/wiki/Systemd-integration) and there should be a few issues/PRs IIRC) but AFAIK there are many(?) NixOS users that want/use it. I think if there are NixOS users interested in maintaining that it might make sense to add an option to the module (e.g. `programs.sway.systemdIntegration`) that is disabled by default but can be quickly enabled (and should then work with the Sway/Wayland related Nixpkgs packages - not sure if they can install systemd user units without a module though). But given that it's not officially supported and we likely add systemd user units that integrate with the sway unit it's likely better suited for home-manager(?).

> If kanshi was working and I knew that GDM was a good choice for most people (SDDM is mentioned originally in this issue) I'd be happy to extend the wiki with a few lines on display manager integration.

That seems like a good idea. The Wiki is community maintained and the current version of the Sway page is AFAIK only a draft to document the systemd integration. So it would be good to extend it. And if there's e.g. an example for GDM that might be easily portable to other login/display managers as well.

**Comment by mschwaig** [comment]
@primeos, thank you for the reply it was really helpful to me to understand what I can expect to work.

**About getting a nice default sway setup to work:**
@primeos's post makes me think it might be better to not use a display manager at all or get a really minimal one working instead of GDM. I have the suspicion that configuring GDM to not do any power-managment would help, but it looks like right now that would also still be quite complicated to configure, I am not sure (does #54150 apply when I try to [disable GDM's power manamgent](https://askubuntu.com/a/1170831/569366)?). I will also try to run kanshi directly from sway and see if that fixes it.

**About using systemd units:**
I have not used home-manager at all yet (I'm going one step at a time), but it seems like home-manager would be a way better fit to optinally integrate specific services as systemd user units instead of starting them from the sway config. Probably having options that move individual parts ot the desktop into user units directly in NixOS is prone to leave you with one foot inside of your `configuration.nix` file and the other one inside your `~/.config/sway/config` file, because things you change in `configuration.nix` will mean you always also have make the corresponding changes in your sway config file correctly. So maybe home-manager would be a better place to do that kind of thing and maybe to do heavier customization in general.

**Comment by erictapen** [comment]
> Probably having options that move individual parts ot the desktop into user units directly in NixOS is prone to leave you with one foot inside of your configuration.nix file and the other one inside your ~/.config/sway/config file

To avoid this situation I declare the content of global `/etc/sway/config` via `environment.etc."sway/config"` and intentionally have no file in `~/.config/sway/`. Works only for single user setups of course.

Regarding the discussion wether to put sway logic into home-manager I'd like to add that this would leave us without the possibility to test sway integration via NixOS tests, like pretty much all X display-manager do and also cage.

Half a year ago I started bringing my Sway systemd user modules upstream, but never finished it. Here is my progress: https://github.com/erictapen/nixpkgs/tree/sway-systemd-integration (heavily WiP). My goal was to make the proposed systemd user integration described in the [Wiki article I wrote](https://nixos.wiki/wiki/Sway) easier for the end user. I'm still using this setup, but felt that it'd be better to wait with the PR until the discussion about where to put sway and how to integrate it into NixOS is solved.

**Comment by primeos** [comment]
> Regarding the discussion wether to put sway logic into home-manager I'd like to add that this would leave us without the possibility to test sway integration via NixOS tests, like pretty much all X display-manager do and also cage.

Not sure about that one, what wouldn't we be able to do? Sway should be fully functional without systemd user units.

> I'm still using this setup, but felt that it'd be better to wait with the PR until the discussion about where to put sway and how to integrate it into NixOS is solved.

I think it would be best to open a dedicated Nixpkgs issue regarding this (with e.g. the "0.kind: question" and "9.needs: community feedback" labels for better visibility and since this will require some discussion). I'm not too optimistic that it will be easy to nicely integrate this into Nixpkgs tbh but I guess it should be possible. I think the main problems are:
- Wayland is very modular and flexible by design which resulted in a very diverse ecosystem (software wise)
- We don't have any abstractions for Wayland compositors in Nixpkgs (like `services.xserver` for X11) and the previous problem makes this more difficult
- AFAIK there's no real consensus if and how we want to handle systemd user units in Nixpkgs

**Comment by mschwaig** [comment]
I have updated [my config (also linked above)](https://gist.github.com/mschwaig/195fe93ed85dea7aaceaf8e1fc6c0e99) to use SDDM instead of GDM. Kanshi still doesn't pick up when outputs are changed on the fly, but this is now a workable setup for me, because I can recover from being left with a black screen easily and every time.

I also wanted to highlight that when I was using GDM I used the [snippet](https://github.com/NixOS/nixpkgs/issues/57602#issuecomment-657762138) that @wkral posted and that itself was working well for avoiding `xserver.enable = true` and looks promising to me.

**Comment by xunam** [comment]
Hello,

I have been experimenting with Sway for a while so I would like to share some feedback. I am currently using NixOS 20.09 and GDM as the login manager. Things mostly work, the one thing that I find problematic is the setup of environment variables.

More specifically, when enabling Sway with no customization, simply by saying `programs.sway.enable = true` in `configuration.nix`, the session is started without reading `/etc/profile` nor `~/.profile`, so Sway itself and anything it launches ignores any session configuration found in those places. This can be problematic when some variables need to be set up, especially `PATH`. Besides, when starting terminals, these profiles are not read either because they are not login shells.

My current way out of this problem is to enable Sway this way:
```
    programs.sway = {
      enable = true;
      extraSessionCommands = ''
        source /etc/profile
        test -f $HOME/.profile && source $HOME/.profile
        export MOZ_ENABLE_WAYLAND=1
        systemctl --user import-environment
      '';
      wrapperFeatures.gtk = true;
    };
```

I also tried using the `systemd` approach of putting files in `~/.config/environment.d/` and it works for all variables except for `PATH` which apparently gets reinitialized before starting the session. Besides, this approach induces some redundancy with `~/.profile` (unless `environment.d` is somehow used also for login shells).

I would be interested if there are better ways to proceed.

For what it's worth, Gnome session on Wayland apparently does read `/etc/profile` and `~/.profile` at some point.

A few other random thoughts while I am at it:

- The startup sequence for Wayland sessions is a mystery. Is there any standard that would specify what should happen when logging into a session through a Wayland compositor, particulary regarding the role of the login manager vs the role of the compositor ?
- The [NixOS wrapper for Sway](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/sway/wrapper.nix) apparently hard-codes the path to `swaybg` in `PATH`. That makes no sense to me. What is so special about `swaybg` that it has to get special treatment this way ? Before fumbling with `extraSessionCommands`, this weird `PATH` was inherited by anything launched from Sway.
- What is the point of the option `programs.sway.extraPackages` ? Are these packages somehow integrated into the Sway setup or are they just installed like those in `environment.systemPackages` ?

**Comment by primeos** [comment]
> The startup sequence for Wayland sessions is a mystery. Is there any standard that would specify what should happen when logging into a session through a Wayland compositor, particulary regarding the role of the login manager vs the role of the compositor?

AFAIK the compositor isn't responsible for reading any startup files but I'm not aware of any official documentation regarding this. Most Sway users likely start it from a virtual terminal so the login shell will have already taken care of it. For display managers the behaviour likely isn't consistent (IIRC each one tends to do it's own thing).

For Nixpkgs we'd need something like https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix for Wayland. I thought there was already a dedicated issue for this but couldn't find one :o

> The NixOS wrapper for Sway apparently hard-codes the path to swaybg in PATH. That makes no sense to me. What is so special about swaybg that it has to get special treatment this way ? Before fumbling with extraSessionCommands, this weird PATH was inherited by anything launched from Sway.

The discussion is in https://github.com/NixOS/nixpkgs/issues/60434. IIRC this was because it's a fixed dependency of Sway and the solution via `PATH` sucked the least.

> What is the point of the option programs.sway.extraPackages ? Are these packages somehow integrated into the Sway setup or are they just installed like those in environment.systemPackages ?

The latter. I guess the main reasons are to make it easy to avoid optional packages by overriding `programs.sway.extraPackages` and to have a dedicated option for packages that are only/mostly relevant for Sway.

**Comment by xunam** [comment]
> AFAIK the compositor isn't responsible for reading any startup files but I'm not aware of any official documentation regarding this. Most Sway users likely start it from a virtual terminal so the login shell will have already taken care of it.

Well, let me be the counter-example here :-)

> For display managers the behaviour likely isn't consistent (IIRC each one tends to do it's own thing).
> 
> For Nixpkgs we'd need something like https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix for Wayland. I thought there was already a dedicated issue for this but couldn't find one :o

That indeed seems to be the way to proceed. If Sway appears among the possible sessions in a graphical display manager, then it is expected to be started from there and not from a login shell in the console, so the thing that is run as a session should behave accordingly. A simple `sway-session` script could do the job: set Wayland-specific environment variables, read profile files, then `exec sway`.

> > The NixOS wrapper for Sway apparently hard-codes the path to swaybg in PATH. That makes no sense to me. What is so special about swaybg that it has to get special treatment this way ? Before fumbling with extraSessionCommands, this weird PATH was inherited by anything launched from Sway.
> 
> The discussion is in #60434. IIRC this was because it's a fixed dependency of Sway and the solution via `PATH` sucked the least.

So I'll comment in that issue…

**Comment by primeos** [comment]
Sway 1.6 changed the default for `$WAYLAND_DISPLAY` which might cause a few regressions, even if you don't hard-code the value (defaults to `wayland-0` when unset). For a quick workaround you might want to add the following to your Sway configuration (especially the 1st command for D-Bus, the second one is only required for some systemd user units):
```
exec dbus-update-activation-environment WAYLAND_DISPLAY
exec systemctl --user import-environment WAYLAND_DISPLAY
```

See https://github.com/NixOS/nixpkgs/issues/119445#issuecomment-820507505 for more details.

**Comment by evils** [comment]
is there a recommended way forward with regards to dbus on wayland?
i'm specifically asking because apparently setting `MOZ_DBUS_REMOTE` in the wrapper for element-desktop (#123777)
while fixing opening links for wayland users, broke an X11 user's ability to open links from element-desktop in (presumably) firefox

requiring every sway user to add a workaround seems like something we should avoid if this can be managed via `programs.sway.enable` (and equivalents for other wayland environments?)
(particularly, i am unwilling to add such a workaround to my system, this should "just work" and i won't add stuff i'll have to remember to remove later)

**Comment by primeos** [comment]
**Some updates:**

@evils: If this is regarding https://github.com/NixOS/nixpkgs/issues/57602#issuecomment-820512097 / https://github.com/NixOS/nixpkgs/issues/119445 then it's fixed now by https://github.com/NixOS/nixpkgs/pull/122605. That only works for new Sway users though. If you've already copied the old `/etc/sway/config` then this requires adding `include /etc/sway/config.d/*`.

With https://github.com/NixOS/nixpkgs/pull/123034 and https://github.com/NixOS/nixpkgs/pull/122602 screen-sharing should also be easier to get to work now. We've also updated the documentation of the Sway module a little bit.

Btw: Quick heads-up that I'm planning to remove `rxvt-unicode` from `sway.extraPackages` (https://github.com/NixOS/nixpkgs/pull/121925) since Alacritty is the new default terminal for Sway for quite some time now.

We're also considering to use a wallpaper in the default configuration for the Sway module but it wasn't decided yet whether to use Sway's default wallpaper (https://github.com/NixOS/nixpkgs/pull/122995) or a NixOS wallpaper (https://github.com/NixOS/nixpkgs/pull/122992). Feedback is welcome.

**Comment by uvNikita** [comment]
> Btw: Quick heads-up that I'm planning to remove rxvt-unicode from sway.extraPackages (#121925) since Alacritty is the new default terminal for Sway for quite some time now.

Just wanted to mention that sway recently switched to `foot` as a default terminal: https://github.com/swaywm/sway/commit/009c58fc9529e284569bf58b73964cb3dc7c5baf

**Comment by evils** [comment]
> If this is regarding ....

this is regarding 123777
as the comment about `WAYLAND_DISPLAY` seems to be dbus related it seemed likely related to me

**Comment by primeos** [comment]
Heads-up: The Nix package was just updated to Sway 1.7.
More details: https://github.com/NixOS/nixpkgs/commit/2e719d1cdab9e81750c19425a1f5c7678b5e73ad (https://github.com/NixOS/nixpkgs/pull/151902)

---

Notable (backward incompatible) changes:
- The default terminal changed from Alacritty to foot

Known issues:
- `swaynag` will crash when Sway 1.6.1 is still running while the Nix
  package (and thus `swaynag`) is already updated to version 1.7.
- The experimental Ozone/Wayland support of Electron apps will be broken
  for a while. Electron version 17 should work but the Chromium fixes
  haven't yet been backported to Electron version 16.

---

The Electron status is tracked in https://github.com/NixOS/nixpkgs/issues/156352.

**Comment by ymatsiuk** [comment]
Does anyone experience the return of firefox popup issue (literally all firefox extensions are broken since the popups don't render) on latest unstable? It started to happen since the last mass rebuild from staging has landed to unstable.

**Comment by t184256** [comment]
@ymatsiuk mine render, never flicker and even have a solid chance of then never going away; willing to send the excess extraneous ones your way =D I'm now on GNOME tho

**Comment by a-kenji** [comment]
Here is more about the Firefox issue: https://github.com/NixOS/nixpkgs/issues/167785. I think upstream is looking at fixing it as well.

**Comment by viperML** [comment]
Today I was hit by the problem of sway not sourcing `/etc/profile`

As most environment variables probably get set by `pam_env`, but `SSH_AUTH_SOCK` is set by a bash snippet run in `/etc/profile`: https://github.com/NixOS/nixpkgs/blob/c43f676c938662072772339be6269226c77b51b8/nixos/modules/programs/ssh.nix#L348-L353 I found it because `vscode` would fail to connect via ssh when launched outside a shell ( https://github.com/microsoft/vscode-remote-release/issues/8043 ), as `SSH_AUTH_SOCK` had not been set.

An ad-hoc solution is the following:
```nix
  programs.sway.extraSessionCommands = ''
    source /etc/profile
  ''
```

My guessings are that:
- X11 window managers do this sourcing in some sort of xsession init script (?)
- Wayland desktop managers exec bash internally (?)

---

And on a side note, should we wrap sway to get this env, or should sway itself exec bash to get this environment?

**Comment by primeos** [comment]
> Today I was hit by the problem of sway not sourcing `/etc/profile`

That's actually the intended behaviour (but not everyone likes it - IIRC this already came up a couple of times from users of graphical login managers).
Luckily it's already documented in https://github.com/swaywm/sway/wiki/Setting-Environmental-Variables now:
> Environment variables are inherited from the process that starts sway. You need to set variables there.

It would probably best to simply document this in https://nixos.wiki/wiki/Sway (just be aware that it's community maintained and not an official NixOS documentation).

**Comment by teutat3s** [comment]
Does anybody have working idle inhibit with firefox (e.g. video playback, video call) running in sway? Not working at all here, current workaround is waybar's idle inhibit module. Not sure where to start debugging this.

**Comment by Dan-Theriault** [comment]
I've had good results from https://github.com/rafaelrc7/wayland-pipewire-idle-inhibit
